### PR TITLE
[Datasets] Add bulk Parquet file reader API

### LINKED ISSUE #22910: [Feature] Ray dataset loading large list of parquet files is extremely slow
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Trying to use ray to run inference on our big dataset. And we are using ray dataset to load the dataset. The read operation is quite slow on 2k files, and increases linearly proportional to number of files, as our goal is to process 100k files. 

Here is a simple code snippet to "reproduce" the "problem".
```
import gcsfs
import ray

fs = gcsfs.GCSFileSystem()
len(file_names)
# 2048
%time ds = ray.data.read_parquet(file_names, filesystem=fs, parallelism=len(file_names))
# 4 minute
```

### Use case

After some poking around, the bottleneck appears to be using [arrow.parquet.ParquetDataset](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/parquet_datasource.py#L92) as the basis for ray.dataset. It's a single machine, sequential operation. And have no much room to user to scale up. 

```
import pyarrow.parquet as pq
len(file_names)
# 2048
%time ds = pq.ParquetDataset(file_names, filesystem=fs)
# Wall time: 2min 55s
```
appears majority of the time is spend on this operation.  There are some room for improving pyarrow by using multiple cpu, but it doesn't help in our case :(  
```
import pyarrow as pa
pa.set_cpu_count(10)
%time ds = pq.ParquetDataset(file_names, filesystem=fs)
# Wall time: 3min 56s
```

We've used apache beam before for loading dataset for inference. Which doesn't have this type of problem, since it's using [arrow.parquet.ParquetFile](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/parquetio.py#L332), and handles the filenames using beam distributed data abstraction. This is a much more efficient/smart approach, as it would process the files in parallel hence adding machine would definitely help.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by pdames**
## Why are these changes needed?

Adds a reader suitable for quickly reading a large number (e.g. 1-100K+) of Parquet files into a Ray Dataset from either local disk or cloud storage.

Latest Benchmarks:
- Environment 1:
  - Cluster: 5 x r5n.8xlarge (160 vCPUs)
  - Storage: S3
  - 2000 Tiny (~4KB) Parquet Files:
    - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 200)
      - Read Dataset Latency: 1.646850564982742
      - Write Dataset Latency: 5.7583091059932485
      - Read Dataset Latency: 2.4654725349973887
      - Write Dataset Latency: 6.635675027035177
      - Read Dataset Latency: 2.460472490056418
      - Write Dataset Latency: 6.1093138850992545
    - Test Case 2 (`read_parquet()` - default parquet metadata provider - parallel 200)
      - Read Dataset Latency: 15.828366486006416
      - Write Dataset Latency: 12.378745757043362
      - Read Dataset Latency: 15.968373686890118
      - Write Dataset Latency: 12.28318011702504
      - Read Dataset Latency: 16.455996707081795
      - Write Dataset Latency: 12.62632444105111
  - 10,000 Tiny (~4KB) Parquet Files:
     - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 160)
       - Read Dataset Latency: 5.766188704059459
       - Write Dataset Latency: 6.506135637057014
      - Test Case 2 (`read_parquet_fast()` - default fast metadata provider - parallel 320)
        - Read Dataset Latency: 3.841439099982381
        - Write Dataset Latency: 6.254379783989862
     - Test Case 3 (`read_parquet_fast()` - default fast metadata provider - parallel 640)
        - Read Dataset Latency: 2.392879810067825
        - Write Dataset Latency: 6.607215114985593
     - Test Case 4 (`read_parquet_fast()` - default fast metadata provider - parallel 1280)
        - Read Dataset Latency: 2.1967075469437987
        - Write Dataset Latency: 9.095152657013386
     - Test Case 5 (`read_parquet_fast()` - default fast metadata provider - parallel 1600)
        - Read Dataset Latency: 2.277380953077227
        - Write Dataset Latency: 9.146176831098273
     - Test Case 6 (`read_parquet_fast()` - default fast metadata provider - parallel 2560)
        - Read Dataset Latency: 2.137215052964166
        - Write Dataset Latency: 14.531858821981587
  - 50,000 Tiny (~4KB) Parquet Files:
    - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 320)
      - Read Dataset Latency: 11.416414194973186
      - Write Dataset Latency: 12.532715738983825
      - Read Dataset Latency: 11.54179179109633
      - Write Dataset Latency: 12.113915671012364
    - Test Case 2 (`read_parquet_fast()` - default fast metadata provider - parallel 640)
      - Read Dataset Latency: 7.514457287965342
      - Write Dataset Latency: 12.104804654954933
      - Read Dataset Latency: 8.027748159016483
      - Write Dataset Latency: 10.249830988934264
    - Test Case 3 (`read_parquet_fast()` - default fast metadata provider - parallel 1280)
      - Read Dataset Latency: 5.188601649017073
      - Write Dataset Latency: 15.750625844928436
      - Read Dataset Latency: 5.264186333981343
      - Write Dataset Latency: 10.911084560095333
  - 600,000 Tiny (~4KB) Parquet Files:
    - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 1280)
      - Read Dataset Latency: 42.07806533499979
      - Write Dataset Latency: 202.45852285499996
- Environment 2:
  - Cluster: 5 x c5n.9xlarge (180 vCPUs)
  - Storage: S3
  - 50,000 Tiny (~4KB) Parquet Files:
    - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 180)
      - Read Dataset Latency: 20.308360855000046
      - Write Dataset Latency: 18.48248909499989
    - Test Case 2 (`read_parquet_fast()` - default fast metadata provider - parallel 360)   
      - Read Dataset Latency: 8.663979919000212
      - Write Dataset Latency: 18.941202082000018
    - Test Case 3 (`read_parquet_fast()` - default fast metadata provider - parallel 720)   
      - Read Dataset Latency: 6.119230162000349
      - Write Dataset Latency: 18.112871180000184
    - Test Case 4 (`read_parquet_fast()` - default fast metadata provider - parallel 1440)   
      - Read Dataset Latency: 4.157292100000177
      - Write Dataset Latency: 18.078249656000025
    - Test Case 5 (`read_parquet_fast()` - default fast metadata provider - parallel 2880)   
      - Read Dataset Latency: 3.098416549999911
      - Write Dataset Latency: 20.54457710700035
  - 600,000 Tiny (~4KB) Parquet Files:
    - Test Case 1 (`read_parquet_fast()` - default fast metadata provider - parallel 1440)
      - Read Dataset Latency: 32.844527049000135
      - Write Dataset Latency: 185.16646222
    - Test Case 2 (`read_parquet_fast()` - default fast metadata provider - parallel 2880)
      - Read Dataset Latency: 18.028459068999837
      - Write Dataset Latency: 181.07565845699992
      
Old Benchmarks:
- Cluster: 5 x r5n.8xlarge
- Storage: S3
- 2000 Tiny (~2KB) Hive-Partitioned Parquet Files:
  - `read_parquet()` not benchmarked due to ParquetDataset merge failure on indexed partition columns: https://issues.apache.org/jira/browse/ARROW-13851
  - Test Case 1 (`read_parquet_fast()` - cached parquet metadata provider - expand paths skipped - no partition filter)
    -  Read Dataset Latency: 3.72136511397548
    -  Write Dataset Latency: 7.245090199983679
    -  Read Dataset Latency: 2.96225255203899
    -  Write Dataset Latency: 6.7958771920530125
    -  Read Dataset Latency: 2.9959744710940868
    -  Write Dataset Latency: 15.850318810902536
    -  Read Dataset Latency: 3.111023452016525
    -  Write Dataset Latency: 10.99442114494741
  - Test Case 2 (`read_parquet_fast()` - cached parquet metadata provider - expand paths skipped -  partition filter keeping 1/7 files - `partition_filter_fn=lambda d: d["day"] == "MO"`)
    - Read Dataset Latency: 1.5806240870151669
    - Write Dataset Latency: 8.724663103889024
    - Read Dataset Latency: 2.334774918970652
    - Write Dataset Latency: 8.310331499087624
  - Test Case 3 (`read_parquet_fast()` - cached parquet metadata provider - expand paths skipped -  partition filter keeping all files - `partition_filter_fn=lambda d: d["day"] != "NA"`)
    - Read Dataset Latency: 3.1833109749713913
    - Write Dataset Latency: 8.957277923938818
    - Read Dataset Latency: 1.950843447004445
    - Write Dataset Latency: 8.899622384924442
    - Read Dataset Latency: 2.891347081051208
    - Write Dataset Latency: 14.80685784202069    
- 2000 Tiny (~4KB) Parquet Files:
  - Test Case 1 (`read_parquet_fast()` - cached parquet metadata provider - expand paths skipped)
      - Read Dataset Latency: 2.8592376470332965
      - Write Dataset Latency: 8.016067871009
      - Read Dataset Latency: 2.98329175892286
      - Write Dataset Latency: 7.4663043189793825
  - Test Case 2 (`read_parquet()` - cached parquet metadata provider - expand paths skipped)
      - Read Dataset Latency: 1.5026919830124825
      - Write Dataset Latency: 17.154331294004805
      - Read Dataset Latency: 1.437328452942893
      - Write Dataset Latency: 21.36053073592484
      - Read Dataset Latency: 1.3846774999983609
      - Write Dataset Latency: 13.24175373907201
  - Test Case 3 (`read_parquet()`)
      - Read Dataset Latency: 18.912850142922252
      - Write Dataset Latency: 12.525802098913118
      - Read Dataset Latency: 17.814308558939956
      - Write Dataset Latency: 13.623326486092992
      - Read Dataset Latency: 15.942234317073599
      - Write Dataset Latency: 13.625373031012714
  - Test Case 4 (`read_parquet_fast()`)
      - Read Dataset Latency: 21.10480642109178
      - Write Dataset Latency: 6.64842400001362
      - Read Dataset Latency: 30.31197776598856
      - Write Dataset Latency: 7.285999955958687

## Related issue number

Closes https://github.com/ray-project/ray/issues/22910

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Very exciting! @clarkzinzow @naijoaix want to help shepherd this?

File: python/ray/data/datasource/partitioning.py
Comment (ericl) [inline python/ray/data/datasource/partitioning.py:38]: Hmm, is this a common partitioning style? If we only support the first (like Spark does) it may simplify the code.

File: python/ray/data/datasource/partitioning.py
Comment (ericl) [inline python/ray/data/datasource/partitioning.py:105]: Nice.

**Review by ericl — COMMENTED** [review]
@pdames this looks pretty good at a high level. Any major items missing beyond tests?

File: python/ray/data/datasource/partitioning.py
Comment (pdames) [reply python/ray/data/datasource/partitioning.py]: It is relatively common, but not as ubiquitous as hive-style partitioning. CloudTrail and Kinesis Firehose are two notable AWS service examples that use directory partitioning, so this is useful for parsing/filtering partitioned datasets that they produce.

**Comment by pdames** [comment]
> @pdames this looks pretty good at a high level. Any major items missing beyond tests?

I just updated the code to a version that passed a small integration test suite parsing and filtering hive-style partitioned datasets exported from Redshift. I think it's feature-complete at this point, so I'll start working on some associated unit tests.

**Comment by pdames** [comment]
Based on the 2K tiny file benchmarks I just ran, the good news is that (1) the partition filter adds little-to-no overhead (and appropriately reduces overall read latency when filtering out files) and (2) the `read_parquet_fast()` implementation lives up to its name on S3 as long as path expansion is skipped. 

The bad news is that, without a custom metadata provider, `read_parquet_fast()` is almost equivalently slow at round-trip I/O and slower than `read_parquet()` due to the overhead of default path expansion. A couple immediate ways to improve the default latency could be to (1) parallelize path expansion via `ray.remote` and (2) apply partition filters eagerly during path expansion (right now they're applied after path expansion).

**Comment by ericl** [comment]
Hmm, so the issue is that if you provide a list of files, we're needing to double check that each file is not a directory? I guess maybe indeed we should make path expansion a flag that we can turn off, maybe even logging a warning if the user is passing a list and has expansion on.

**Comment by ericl** [comment]
Hmm, so the issue is that if you provide a list of files, we're needing to double check that each file is not a directory? I guess maybe indeed we should make path expansion a flag that we can turn off, maybe even logging a warning if the user is passing a list and has expansion on.

**Comment by ericl** [comment]
Hmm, so the issue is that if you provide a list of files, we're needing to double check that each file is not a directory? I guess maybe indeed we should make path expansion a flag that we can turn off, maybe even logging a warning if the user is passing a list and has expansion on.

**Comment by pdames** [comment]
So, instead of providing a boolean flag to skip path expansion, I created a default `FastFileMetadataProvider` for `read_parquet_fast()` that skips path expansion, and logged associated warnings under both the `DefaultFileMetadataProvider` about high-latency metadata resolution, and no metadata resolution or path expansion under the `FastFileMetadataProvider`. 

The end results are looking pretty good here, with ~80% performance improvement for 2K files vs `read_parquet()`, and with `read_parquet_fast()` completing in ~18 seconds for 600K tiny S3 files (and a little over 3 minutes for a round trip read/write @ parallel 2880 on a 5 x c5n.9xlarge cluster).

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [inline python/ray/data/datasource/file_based_datasource.py:223]: I'm not sure we should log this unconditionally: the input paths may contain directory so it's not appropriate to suggest FastFileMetadataProvider. How about commenting the class that if users are dealing with only files, they should use FastFileMetadataProvider?

**Review by jianoaix — COMMENTED** [review]
Thank you Patric for making this improvement!

File: python/ray/data/datasource/file_based_datasource.py
Comment (ericl) [inline python/ray/data/datasource/file_based_datasource.py:223]:
```suggestion
       if len(paths) > 1:
           logger.warning(
```

This should hide the warning for the common case, where a single directory is passed as the path.

File: python/ray/data/read_api.py
Comment (pdames) [reply python/ray/data/read_api.py:378]: An OSError (identical to the OSError for a non-existent file path) is raised currently - I've added this to the documentation.

File: python/ray/data/read_api.py
Comment (pdames) [reply python/ray/data/read_api.py:386]: Yes, this uses the base `prepare_read()` method from `FileBasedDatasource`, so the schema remains undefined by default (just like with CSV/NumPy/JSON/Binary datasources). If known, the schema can be added by either specifying it as a keyword argument and/or via a custom metadata provider.

File: python/ray/data/datasource/partitioning.py
Comment (pdames) [reply python/ray/data/datasource/partitioning.py:216]: Removed!

File: python/ray/data/datasource/partitioning.py
Comment (pdames) [reply python/ray/data/datasource/partitioning.py:36]: My thinking here is that this class is really a path-based partition parser that currently only exposes a filter to end-users. Ideally, it should be extended with methods to cover additional tasks in subsequent PRs like (1) `group(paths: List[str]) -> Dict[Tuple[str, str], List[str]]` to group paths by partition (e.g. to facilitate a partitioned dataset read), (2) `partition_dir(base_dir: str, partition_values: List[str]) -> str` to find the correct directory for a list of partition values (e.g. for a partitioned dataset write), (3) `convert(paths: List[str], from: PartitionStyle, to: PartitionStyle)` to convert a list of paths between partition styles, etc.

File: python/ray/data/datasource/parquet_fast_datasource.py
Comment (pdames) [reply python/ray/data/datasource/parquet_fast_datasource.py:51]: I had some similar thoughts. However, the class itself supports directories, so I'm not sure about either `Directoryless` or `DirectFile`. It's just the `FastFileMetadataProvider` default in the `read_parquet_fast()` API that alters it to not handle directories.

`ParquetNoMetaDatasource` was one thing I was thinking of, since not overriding `prepare_read()` means it's not populated with verbose metadata, but I can also imagine this changing in the future.

I think it may be best to just call this something like `ParquetBasicDatasource` to indicate that it is a minimal implementation that reuses as much code as possible from `FileBasedDatasource`. 

As far as `read_parquet_fast()` API naming goes, I've been thinking that `read_parquet_bulk()` would be more accurate, since the current implementation offers little to no performance benefits when reading a small number of files (esp. if each file is relatively large), but I expect it to continue to offer long-term benefits for bulk reads of a large number of files.

Comment (pdames) [reply python/ray/data/datasource/parquet_fast_datasource.py]: My last commit changes `read_parquet_fast()` to `read_parquet_bulk()`, `ParquetFastDatasource` to `ParquetBaseDatasource`, and then has `ParquetDatasource` inherit from `ParquetBaseDatasource` since they share the same file format and write path.

What do you think?

File: python/ray/data/datasource/file_based_datasource.py
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py]: Another option that I was considering here is to only log this warning after a fixed amount of time has elapsed expanding directory paths (e.g. if we've been expanding directories for more than 15 seconds, then log this warning).

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [inline python/ray/data/datasource/file_based_datasource.py:223]: Why is this a good indicator to log? The user can supply a mix of files and directories in the 'paths', right?

File: python/ray/data/datasource/partitioning.py
Comment (jianoaix) [reply python/ray/data/datasource/partitioning.py:36]: Agree this is extendable. Shall we name it so, e.g. PathPartitionParser, or PathPartitionProcessor?

**Review by jianoaix — COMMENTED** [review]
Instead of adding a big PR and then a follow-up big unit test PR, would it be possible to shard this PR horizontally and add in corresponding unit tests for each smaller PRs?
E.g. 1) partitioning.py etc.; 2) parque_base_datasource.py etc; 3) integration into read_api.py.

File: python/ray/data/datasource/file_based_datasource.py
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py:223]: The assumption here is that the most common use-case is that (1) a single directory is passed containing input files or (2) multiple paths are passed containing input files. Mixed file and directory inputs are assumed to be less common. However, I think I could also change this to only log the warning after a fixed amount of time has elapsed expanding directories, which may be more generally appropriate.

**Comment by pdames** [comment]
> Instead of adding a big PR and then a follow-up big unit test PR, would it be possible to shard this PR horizontally and add in corresponding unit tests for each smaller PRs? E.g. 1) partitioning.py etc.; 2) parque_base_datasource.py etc; 3) integration into read_api.py.

Sounds good to me. Now that we understand the total scope of changes from this PR, I can start opening up smaller PRs and linking them back to this PR and the parent issue that we're working on. I think the proposed ordering of PRs given above also makes sense.

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py:223]: Ok, it seems fine as it's simpler (since it's just a precaution).

File: python/ray/data/datasource/parquet_fast_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/parquet_fast_datasource.py]: Looks good.

**Review by jianoaix — COMMENTED** [review]
Thank you for addressing comments, LGTM once it's split into smaller PRs with unit tests.

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py]: Ok, it seems fine to keep this simple since this is a precaution.

File: python/ray/data/datasource/partitioning.py
Comment (pdames) [reply python/ray/data/datasource/partitioning.py:36]: `PathPartitionParser` sounds good to me.

**Comment by pdames** [comment]
Rebased on top of https://github.com/ray-project/ray/pull/23624 to ensure end-to-end compatibility of that PR with upcoming PRs. No need to review here.

File: python/ray/data/tests/test_dataset_formats.py
Comment (jianoaix) [inline python/ray/data/tests/test_dataset_formats.py:434]: Add a test for reading directory with right metadata provider, so it should have no issue?

**Review by jianoaix — COMMENTED** [review]
The delta part looks good. Will take another look after second PR merged.

File: python/ray/data/tests/test_dataset_formats.py
Comment (pdames) [reply python/ray/data/tests/test_dataset_formats.py:435]: Added `test_parquet_read_bulk_meta_provider` for this in my latest commit.

**Comment by pdames** [comment]
@jianoaix @clarkzinzow Rebased on top of the latest from master and ready for final review.

**Review by clarkzinzow — APPROVED** [review]
LGTM!

File: python/ray/data/datasource/file_based_datasource.py
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py]: Nope! This would be a good way to minimize the footprint of changes, so I'll push that as a quick update.

Comment (clarkzinzow) [reply python/ray/data/datasource/file_based_datasource.py]: Ah @pdames what do you think about doing this as a follow-up? We're racing against the branch cut clock right now (4pm PT), don't want to risk having to iterate on a few CI builds! 😉

Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py]: 🤞

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/read_api.py
Comment (clarkzinzow) [reply python/ray/data/read_api.py:333]: @jianoaix I believe that column projection will happen via Arrow's `pq.read_table()` [API](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html), where the column selection is passed through to that call when using this bulk Parquet API.

**Comment by clarkzinzow** [comment]
Datasets tests look good, merging! Great work @pdames! 🙌

**Pull request merged**
